Add initialization from Twiss parameters

n01r opened the pull request:
This changes the initialization of the beam particle distribution to allow for Twiss parameters to be supplied.
The code first checks if the name for the supplied distribution has a suffix indicated by an underscore. It then checks if the base name of the distribution is an element of the list of allowed distributions that share the same input logic. Then it chooses to initialize the beam ellipse parameters either directly from the given inputs, or it calculates them from the Twiss parameters and proceeds with initialization.

Closes #406.

Tasks:
- [x] implementation
- [x] documentation
  - [x] emphasize that unnormalized emittance is the input and/or provide conversion via reference beta-gamma
- [x] dedicated test
- [x] Python bindings $\Longrightarrow$ follow-up PR #697

n01r committed:
Add initialization from Twiss parameters

This changes the initialization of the beam particle distribution to allow for
Twiss parameters to be supplied.
The code first checks if the name for the supplied distribution has a suffix
indicated by an underscore. It then checks if the base name of the distribution
is an element of the list of allowed distributions that share the same input
logic. Then it chooses to initialize the beam ellipse parameters either
directly from the given inputs, or it calculates them from the Twiss parameters
and process with initialization.

cemitch99 reviewed (commented):
The math here looks good to me.  We should specify in the docs that the unnormalized emittances are used as input (or add conversion by the reference beta*gamma).

n01r commented:
Thanks for checking, @cemitch99! Right, I added the topic to the meeting agenda today but I will also add some tasks to this PR, like docs etc. :)

ax3l commented on src/initialization/InitDistribution.H:
```diff
@@ -98,6 +98,26 @@ namespace impactx::initialization
     };
+
+    /** Initialize the input parameters for all distributions that read phase space ellipse parameters from the input
+     *
```
Please add `@param[in] name description here` doxygen strings on the two new function declarations here.

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -121,6 +123,91 @@ namespace impactx
+            amrex::Abort("Input Error: Emittance values need to be non-zero positive values in all dimensions.");
+        }
+
+        amrex::Vector<amrex::ParticleReal> alphas = {alphax, alphay, alphat};
+        amrex::Vector<amrex::ParticleReal> betas = {betax, betay, betat};
+        amrex::Vector<amrex::ParticleReal> emittances = {emittx, emitty, emittt};
```
I think these can be `std::array<amrex::ParticleReal, 3> const`?

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -22,8 +22,10 @@
+#include <cmath>
 #include <string>
 #include <type_traits>
+#include <tuple>
```
Please keep stdlib includes ordered alphabetically :)

n01r replied:
I will remove this include actually because my current implementation does not make use of `std::tuple` anymore.

n01r committed:
Add documentation and tests

Added two tests, one for distribution generation and one for the simple
FODO cell that use the new input from Twiss parameters.
Added documentation on our units and the two beam input systems.
Added a phase space ellipse figure.

ax3l commented on docs/source/theory/intro.rst:
```diff
@@ -55,6 +55,38 @@ Unlike particles within the beam, the reference particle is described by a set o
+Beam Distribution Input
+-----------------------
+
+.. figure:: phase_space_ellipse.svg
```
With 1.1kB, this svg seems small enough to check in (instead of adding to the PR description)

ax3l commented on src/initialization/InitDistribution.H:
```diff
@@ -98,6 +98,58 @@ namespace impactx::initialization
+     * @param mutpt Reference to the variable storing the mean for energy.
+     */
+    void
+    setDistributionParametersFromPhaseSpaceInputs (
```
Feel free to use `set_distribution_...` spelling for functions/methods in ImpactX. I started a bit inconsistently, but we want to use that going forward. Also makes it nicely the same in Python bindings.

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -121,6 +125,91 @@ namespace impactx
+        pp_dist.get("emittT", emittt);
+
+        if (betax <= 0.0_prt || betay <= 0.0_prt || betat <= 0.0_prt) {
+            amrex::Abort("Input Error: The beta function values need to be non-zero positive values in all dimensions.");
```
Please throw a runtime exception instead. This is input, so a user might want to recover in a Python notebook from it.

AMReX' abort does something more complex: https://amrex-codes.github.io/amrex/docs_html/Basics.html?highlight=amrex+abort#abort-assertion-and-backtrace

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -121,6 +125,91 @@ namespace impactx
+        }
+
+        if (emittx <= 0.0_prt || emitty <= 0.0_prt || emittt <= 0.0_prt) {
+            amrex::Abort("Input Error: Emittance values need to be non-zero positive values in all dimensions.");
```
See above.

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -171,163 +260,144 @@ namespace impactx
-          pp_dist.get("sigmaPt", sigpt);
-          pp_dist.query("muxpx", muxpx);
-          pp_dist.query("muypy", muypy);
-          pp_dist.query("mutpt", mutpt);
-
-          distribution::KnownDistributions const triangle(distribution::Triangle(
-            sigx, sigy, sigt,
-            sigpx, sigpy, sigpt,
-            muxpx, muypy, mutpt));
-
-          add_particles(bunch_charge, triangle, npart);
+        std::string base_dist_type = distribution_type;
```
The "has `_`" is not as robust as I would prefer. Can we instead change the logic to "ends in `_from_cs` || `_from_twiss`"?

ax3l reviewed (commented):
This looks great! Small comments added for improvement suggestions :)

n01r committed:
Implement reviewer comments

- change logic from "has underscore" to "ends in `_from_twiss` or `_from_cs`"
  for distribution name
- replace `amrex::Abort` with `throw std::runtime_error` to catch it in Python
- changed naming of `setDistributionParameters...`
  to `set_distribution_parameters...` to make it more similar to Python methods

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -22,13 +22,17 @@
 #include <AMReX_Print.H>
 
+#include <cmath>
 #include <string>
```
For `std::runtime_error`:
```suggestion
#include <stdexcept>
#include <string>
```

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -121,6 +125,91 @@ namespace impactx
+        amrex::ParticleReal& sigpx, amrex::ParticleReal& sigpy, amrex::ParticleReal& sigpt,
+        amrex::ParticleReal& muxpx, amrex::ParticleReal& muypy, amrex::ParticleReal& mutpt
+    )
+    {
+
```
```suggestion
    {
        using namespace amrex::literals; // for _rt and _prt

```

ax3l replied:
```suggestion
```

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -22,13 +22,17 @@
 
 namespace impactx
 {
+
+    using namespace amrex::literals; // for _rt and _prt
+
```
Let's scope this more, not needed in default arguments, no?
```suggestion
```

ax3l replied:
I do not see where it is needed, but if you do need it somewhere please keep `using namespace` inside function bodies whenever possible.

n01r replied:
I am using this to have the exact same types in my comparison statements.
It's probably overkill
```C++
if (betax <= 0.0_prt || betay <= 0.0_prt || betat <= 0.0_prt) {
    throw std::runtime_error("Input Error: The beta function values need to be non-zero positive values in all dimensions.");
}
```

n01r replied:
I moved it inside the function body now :+1:

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -121,6 +125,91 @@ namespace impactx
+        amrex::ParticleReal& muxpx, amrex::ParticleReal& muypy, amrex::ParticleReal& mutpt
+    )
+    {
+
```
```suggestion
```

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -121,6 +125,91 @@ namespace impactx
+        // calculate Twiss / Courant-Snyder gammas
+        amrex::Vector<amrex::ParticleReal> gammas;
+        for (size_t i = 0; i < alphas.size(); i++)
+            gammas.push_back((1.0 + pow(alphas.at(i), 2)) / betas.at(i));
```
You want this: https://en.cppreference.com/w/cpp/numeric/math/pow
```suggestion
            gammas.push_back((1.0 + std::pow(alphas.at(i), 2)) / betas.at(i));
```

Otherwise it is the C function for `double`:
https://en.cppreference.com/w/c/numeric/math/pow

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -121,6 +125,91 @@ namespace impactx
+        muxpx = correlations.at(0);
+        muypy = correlations.at(1);
+        mutpt = correlations.at(2);
+
```
```suggestion
```

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -121,6 +125,91 @@ namespace impactx
+
+        // calculate intersections of phase space ellipse with coordinate axes and the correlation factors
+        for (size_t k = 0; k < betas.size(); k++){
+            sigmas_pos.push_back(sqrt(emittances.at(k)/gammas.at(k)));
+            sigmas_mom.push_back(sqrt(emittances.at(k)/betas.at(k)));
```
```suggestion
            sigmas_pos.push_back(std::sqrt(emittances.at(k)/gammas.at(k)));
            sigmas_mom.push_back(std::sqrt(emittances.at(k)/betas.at(k)));
```

Same as for pow :)

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -121,6 +125,91 @@ namespace impactx
+            sigmas_pos.push_back(sqrt(emittances.at(k)/gammas.at(k)));
+            sigmas_mom.push_back(sqrt(emittances.at(k)/betas.at(k)));
+
+            correlations.push_back(alphas.at(k) / sqrt(betas.at(k) * gammas.at(k)));
```
```suggestion
            correlations.push_back(alphas.at(k) / std::sqrt(betas.at(k) * gammas.at(k)));
```

n01r committed:
Implement more reviewer comments

- remove some newlines
- add `std::` to math functions where I overlooked it

n01r commented:
Thanks for your review!
New changes are pushed.
